File: Backend/src/main/java/lan/dk/podcastserver/manager/worker/downloader/HTTPDownloader.java
package lan.dk.podcastserver.manager.worker.downloader;

import com.github.axet.wget.WGet;
import com.github.axet.wget.info.DownloadInfo;
import com.github.axet.wget.info.ex.DownloadInterruptedError;
import com.github.axet.wget.info.ex.DownloadMultipartError;
import lan.dk.podcastserver.entity.Item;
import lan.dk.podcastserver.manager.ItemDownloadManager;
import lan.dk.podcastserver.service.UrlService;
import lan.dk.podcastserver.service.factory.WGetFactory;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.io.FilenameUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.util.Objects;

import static java.util.Objects.isNull;
import static java.util.Objects.nonNull;

@Slf4j
@Scope("prototype")
@Component("HTTPDownloader")
public class HTTPDownloader extends AbstractDownloader {

    @Autowired
    UrlService urlService;
    @Autowired WGetFactory wGetFactory;

    DownloadInfo info = null;

    private final HTTPWatcher itemSynchronisation = new HTTPWatcher(this);

    @Override
    public Item download() {
        log.debug("Download");

        try {
            info = wGetFactory.newDownloadInfo(urlService.getRealURL(getItemUrl(item)));
            info.extract(stopDownloading, itemSynchronisation);
            target = getTargetFile(item);
            WGet w = wGetFactory.newWGet(info, target.toFile());
            w.download(stopDownloading, itemSynchronisation);
        } catch (DownloadMultipartError e) {
            e.getInfo().getParts()
                .stream()
                .map(DownloadInfo.Part::getException)
                .filter(Objects::nonNull)
                .forEach(Throwable::printStackTrace);

            stopDownload();
        } catch (DownloadInterruptedError e) {
            log.debug("Arrêt du téléchargement");
        } catch (IOException e) {
            log.debug("Exception during download", e);
            stopDownload();
        }
        return item;
    }

    @Override
    public Integer compatibility(String url) {
        return url.startsWith("http") ? Integer.MAX_VALUE-1 : Integer.MAX_VALUE;
    }

    @Slf4j
    static class HTTPWatcher implements Runnable {

        final HTTPDownloader httpDownloader;

        HTTPWatcher(HTTPDownloader httpDownloader) {
            this.httpDownloader = httpDownloader;
        }

        @Override
        public void run() {

            DownloadInfo info = httpDownloader.info;
            ItemDownloadManager itemDownloadManager = httpDownloader.itemDownloadManager;
            Item item = httpDownloader.item;

            switch (info.getState()) {
                case EXTRACTING:
                case EXTRACTING_DONE:
                    log.debug(FilenameUtils.getName(String.valueOf(httpDownloader.getItemUrl(item))) + " " + info.getState());
                    break;
                case ERROR:
                    httpDownloader.stopDownload();
                    break;
                case DONE:
                    log.debug(FilenameUtils.getName(String.valueOf(httpDownloader.getItemUrl(item))) + " - Téléchargement terminé");
                    httpDownloader.finishDownload();
                    itemDownloadManager.removeACurrentDownload(item);
                    break;
                case RETRYING:
                    log.debug(FilenameUtils.getName(String.valueOf(httpDownloader.getItemUrl(item))) + " " + info.getState() + " " + info.getDelay());
                    break;
                case DOWNLOADING:
                    if (isNull(info.getLength()) || (nonNull(info.getLength()) && info.getLength() == 0L)) break;

                    int progression = (int) (info.getCount()*100 / (float) info.getLength());
                    if (item.getProgression() < progression) {
                        item.setProgression(progression);
                        log.debug("Progression de {} : {}%", item.getTitle(), progression);
                        httpDownloader.convertAndSaveBroadcast();
                    }
                    break;
                case STOP:
                    log.debug("Pause / Arrêt du téléchargement du téléchargement");
                    break;
                default:
                    break;
            }
        }
    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/manager/downloader/HTTPDownloader.kt
package com.github.davinkevin.podcastserver.manager.downloader

import com.github.axet.wget.info.DownloadInfo
import com.github.axet.wget.info.URLInfo.States.*
import com.github.axet.wget.info.ex.DownloadInterruptedError
import com.github.axet.wget.info.ex.DownloadMultipartError
import com.github.davinkevin.podcastserver.download.DownloadRepository
import com.github.davinkevin.podcastserver.messaging.MessagingTemplate
import com.github.davinkevin.podcastserver.service.MimeTypeService
import com.github.davinkevin.podcastserver.service.UrlService
import com.github.davinkevin.podcastserver.service.factory.WGetFactory
import com.github.davinkevin.podcastserver.service.properties.PodcastServerParameters
import org.apache.commons.io.FilenameUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.config.ConfigurableBeanFactory.SCOPE_PROTOTYPE
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component
import java.time.Clock
import java.util.concurrent.atomic.AtomicBoolean

@Suppress("SpringJavaInjectionPointsAutowiringInspection")
@Scope(SCOPE_PROTOTYPE)
@Component
class HTTPDownloader(
        downloadRepository: DownloadRepository,
        podcastServerParameters: PodcastServerParameters,
        template: MessagingTemplate,
        mimeTypeService: MimeTypeService,
        clock: Clock,
        val urlService: UrlService,
        val wGetFactory: WGetFactory
) : AbstractDownloader(downloadRepository, podcastServerParameters, template, mimeTypeService, clock) {

    private val log = LoggerFactory.getLogger(HTTPDownloader::class.java)

    private val stopDownloading = AtomicBoolean(false)
    internal lateinit var info: DownloadInfo

    override fun download(): DownloadingItem {
        log.debug("Download")
        try {
            val url = downloadingInformation.url()
            val realUrl = urlService.getRealURL(url)

            info = wGetFactory.newDownloadInfo(realUrl)

            val itemSynchronisation = HTTPWatcher(this)

            info.extract(stopDownloading, itemSynchronisation)

            target = computeTargetFile(downloadingInformation)
            val w = wGetFactory.newWGet(info, target!!.toFile())

            w.download(stopDownloading, itemSynchronisation)
        } catch (e: DownloadMultipartError) {
            e.info.parts
                    .stream()
                    .map { it.exception }
                    .filter { it != null }
                    .forEach { it.printStackTrace() }

            throw RuntimeException(e)
        } catch (e: DownloadInterruptedError) {
            log.debug("Arrêt du téléchargement")
        }

        return downloadingInformation.item
    }

    override fun startDownload() {
        stopDownloading.set(false)
        super.startDownload()
    }

    override fun pauseDownload() {
        stopDownloading.set(true)
        super.pauseDownload()
    }

    override fun stopDownload() {
        stopDownloading.set(true)
        super.stopDownload()
    }

    override fun failDownload() {
        stopDownloading.set(true)
        super.failDownload()
    }

    override fun compatibility(downloadingInformation: DownloadingInformation) =
            if (downloadingInformation.urls.size == 1 && downloadingInformation.urls.first().startsWith("http")) Integer.MAX_VALUE - 1
            else Integer.MAX_VALUE

    internal fun progression(progress: Int) {
        log.debug("Progression of {} : {}%", downloadingInformation.item.title, progress)
        downloadingInformation = downloadingInformation.progression(progress)
        broadcast(downloadingInformation.item)

    }
}

class HTTPWatcher(private val downloader: HTTPDownloader) : Runnable {

    val log = LoggerFactory.getLogger(HTTPWatcher::class.java)!!

    override fun run() {

        val id = downloader.downloadingInformation.item.id

        val info = downloader.info
        val itemDownloadManager = downloader.itemDownloadManager
        val dInfo = downloader.downloadingInformation
        val fileName = FilenameUtils.getName(downloader.downloadingInformation.url())

        when (info.state) {
            EXTRACTING, EXTRACTING_DONE -> log.debug("$fileName ${info.state}")
            DONE -> {
                log.debug("$fileName - download finished")
                downloader.finishDownload()
                itemDownloadManager.removeACurrentDownload(id)
            }
            RETRYING -> log.debug("$fileName ${info.state} ${info.delay}")
            DOWNLOADING -> {
                if (info.length == null || info.length == 0L) {
                    return
                }

                val progression = (info.count * 100 / info.length.toFloat()).toInt()
                if (dInfo.item.progression < progression) {
                    downloader.progression(progression)
                }
            }
            STOP -> log.debug("Pause / Stop of the download")
            else -> {}
        }
    }
}

